File: CarHomeWidget/src/com/anod/car/home/UtilitiesBitmap.java
package com.anod.car.home;

import java.io.ByteArrayOutputStream;
import java.io.IOException;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.ColorMatrixColorFilter;
import android.graphics.LightingColorFilter;
import android.graphics.Paint;
import android.graphics.PaintFlagsDrawFilter;
import android.graphics.Rect;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.PaintDrawable;
import android.util.Log;

public class UtilitiesBitmap {

    private static int sIconWidth = -1;
    private static int sIconHeight = -1;
    
    private static final Rect sOldBounds = new Rect();
    private static final Canvas sCanvas = new Canvas();
    private static final ColorMatrixColorFilter sGreyColorMatrixColorFilter ;
        
    static {
        sCanvas.setDrawFilter(new PaintFlagsDrawFilter(Paint.DITHER_FLAG,
                Paint.FILTER_BITMAP_FLAG));
        float[] greyScaleMatrix = new float[] { 
        		0.5f, 0.5f, 0.5f, 0, 0, //red
        		0.5f, 0.5f, 0.5f, 0, 0, //green
        		0.5f, 0.5f, 0.5f, 0, 0, //blue
        		0, 0, 0, 1.0f, 0 //alpha
        };
        sGreyColorMatrixColorFilter = new ColorMatrixColorFilter(greyScaleMatrix);
    }
    
    private static void initStatics(Context paramContext)
    {
      int i = (int)paramContext.getResources().getDimension(R.dimen.icon_size);
      sIconHeight = i;
      sIconWidth = i;
      PaintFlagsDrawFilter localPaintFlagsDrawFilter = new PaintFlagsDrawFilter(4, 2);
      sCanvas.setDrawFilter(localPaintFlagsDrawFilter);
    }
 
    /**
     * Returns a bitmap suitable for the all apps view.  The bitmap will be a power
     * of two sized ARGB_8888 bitmap that can be used as a gl texture.
     */
    static Bitmap createIconBitmap(Drawable icon, Context context) {
        synchronized (sCanvas) { // we share the statics :-(
            if (sIconWidth == -1) {
                initStatics(context);
            }

            int width = sIconWidth;
            int height = sIconHeight;

            if (icon instanceof PaintDrawable) {
                PaintDrawable painter = (PaintDrawable) icon;
                painter.setIntrinsicWidth(width);
                painter.setIntrinsicHeight(height);
            } else if (icon instanceof BitmapDrawable) {
                // Ensure the bitmap has a density.
                BitmapDrawable bitmapDrawable = (BitmapDrawable) icon;
                Bitmap bitmap = bitmapDrawable.getBitmap();
                if (bitmap.getDensity() == Bitmap.DENSITY_NONE) {
                    bitmapDrawable.setTargetDensity(context.getResources().getDisplayMetrics());
                }
            }
            int sourceWidth = icon.getIntrinsicWidth();
            int sourceHeight = icon.getIntrinsicHeight();

            if (sourceWidth > 0 && sourceWidth > 0) {
                // There are intrinsic sizes.
                if (width < sourceWidth || height < sourceHeight) {
                    // It's too big, scale it down.
                    final float ratio = (float) sourceWidth / sourceHeight;
                    if (sourceWidth > sourceHeight) {
                        height = (int) (width / ratio);
                    } else if (sourceHeight > sourceWidth) {
                        width = (int) (height * ratio);
                    }
                } else if (sourceWidth < width && sourceHeight < height) {
                    // It's small, use the size they gave us.
                    width = sourceWidth;
                    height = sourceHeight;
                }
            }

            final Bitmap bitmap = Bitmap.createBitmap(sIconWidth, sIconHeight,
                    Bitmap.Config.ARGB_8888);
            final Canvas canvas = sCanvas;
            canvas.setBitmap(bitmap);

            final int left = (sIconWidth-width) / 2;
            final int top = (sIconHeight-height) / 2;

            sOldBounds.set(icon.getBounds());
            icon.setBounds(left, top, left+width, top+height);
           	icon.setColorFilter(null);
            icon.draw(canvas);
            icon.setBounds(sOldBounds);
            return bitmap;
        }
    }

 	static Bitmap scaleBitmap(Bitmap icon, float scale, Context context) {
        if (sIconWidth == -1) {
            initStatics(context);
        }
 		int scW = (int)(sIconWidth*scale);
 		int scH = (int)(sIconHeight*scale);
       	return Bitmap.createScaledBitmap(icon,scW ,scH, true);   	
    }
    static Bitmap applyBitmapFilter(Bitmap icon, Context context) {
        if (sIconWidth == -1) {
            initStatics(context);
        }
        final Bitmap bitmap = Bitmap.createBitmap(sIconWidth, sIconHeight,
                Bitmap.Config.ARGB_8888);
        final Canvas canvas = sCanvas;
        canvas.setBitmap(bitmap);
		
        BitmapDrawable d = new BitmapDrawable(context.getResources(), icon);
        sOldBounds.set(d.getBounds());
        d.setBounds(0, 0, sIconWidth, sIconHeight);
       	d.setColorFilter(sGreyColorMatrixColorFilter);
       	d.draw(canvas);       
        d.setBounds(sOldBounds);
		return bitmap;
    }
    
    static Bitmap tint(Bitmap icon, int color) {
    	Paint p = new Paint(color);
    	LightingColorFilter filter = new LightingColorFilter(color, 0);
    	p.setColorFilter(filter);

        final Canvas canvas = sCanvas;
        canvas.drawBitmap(icon, 0, 0, p);
        return icon;
    }

    
    static byte[] flattenBitmap(Bitmap bitmap) {

        // Try go guesstimate how much space the icon will take when serialized
        // to avoid unnecessary allocations/copies during the write.
        int size = bitmap.getWidth() * bitmap.getHeight() * 4;
        ByteArrayOutputStream out = new ByteArrayOutputStream(size);
        try {
            bitmap.compress(Bitmap.CompressFormat.PNG, 100, out);
            out.flush();
            out.close();
            return out.toByteArray();
        } catch (IOException e) {
            Log.w("Favorite", "Could not write icon");
            return null;
        }
    }    
}

File: app/src/main/java/com/anod/car/home/utils/UtilitiesBitmap.kt
package com.anod.car.home.utils

import android.content.Context
import android.content.pm.PackageManager
import android.graphics.*
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable
import android.graphics.drawable.PaintDrawable
import android.util.DisplayMetrics
import android.util.Log
import com.anod.car.home.R
import java.io.ByteArrayOutputStream
import java.io.IOException
import kotlin.math.max

object UtilitiesBitmap {
    private const val sizeIcon = 0 // icon with higher density
    private const val sizeSystem = 1 // default system icon size
    private const val sizeMax = 2 // max scale size
    const val maxScale = 3

    private var sIconSize = -1
    private var sIconSystem = -1
    private var sIconMaxScale = -1
    private var sSystemDensity = -1
    private var sIconDensity = -1

    private val sOldBounds = Rect()
    private val sCanvas = Canvas()

    init {
        sCanvas.drawFilter = PaintFlagsDrawFilter(Paint.DITHER_FLAG, Paint.FILTER_BITMAP_FLAG)
    }

    private fun initStatics(paramContext: Context) {
        val r = paramContext.resources

        val metrics = r.displayMetrics
        sSystemDensity = metrics.densityDpi
        sIconDensity = getHigherDensity(sSystemDensity)

        sIconSystem = r.getDimension(R.dimen.icon_size).toInt()
        sIconSize = convertForDensity(sIconSystem, sIconDensity, metrics.densityDpi).toInt()

        sIconMaxScale = sIconSize * maxScale

        val localPaintFlagsDrawFilter = PaintFlagsDrawFilter(4, 2)
        sCanvas.drawFilter = localPaintFlagsDrawFilter
    }

    private fun convertForDensity(value: Int, densityDpi: Int, deviceDensityDpi: Int): Float {
        return if (deviceDensityDpi == densityDpi) {
            value.toFloat()
        } else densityDpi / deviceDensityDpi.toFloat() * value
    }

    fun getSystemIconSize(context: Context): Int {
        synchronized(sCanvas) {
            // we share the statics :-(
            if (sIconSize == -1) {
                initStatics(context)
            }

            return sIconSystem
        }
    }

    fun getIconMaxSize(context: Context): Int {
        synchronized(sCanvas) {
            // we share the statics :-(
            if (sIconMaxScale == -1) {
                initStatics(context)
            }

            return sIconMaxScale
        }
    }

    fun getTargetDensity(context: Context): Int {
        synchronized(sCanvas) {
            // we share the statics :-(
            if (sIconSize == -1) {
                initStatics(context)
            }

            return sIconDensity
        }
    }

    private fun getHigherDensity(deviceDensity: Int): Int {

        if (deviceDensity <= DisplayMetrics.DENSITY_TV) {
            return DisplayMetrics.DENSITY_XXHIGH
        } else if (deviceDensity < DisplayMetrics.DENSITY_XXXHIGH) {
            return DisplayMetrics.DENSITY_XXXHIGH
        }

        return deviceDensity
    }

    fun makeDefaultIcon(manager: PackageManager): Bitmap {
        val d = manager.defaultActivityIcon
        val b = Bitmap.createBitmap(max(d.intrinsicWidth, 1),
                max(d.intrinsicHeight, 1), Bitmap.Config.ARGB_8888)
        val c = Canvas(b)
        d.setBounds(0, 0, b.width, b.height)
        d.draw(c)
        return b
    }

    fun createSystemIconBitmap(icon: Drawable, context: Context): Bitmap {
        return createIconBitmapSize(icon, sizeSystem, context)
    }

    fun createHiResIconBitmap(icon: Drawable, context: Context): Bitmap {
        return createIconBitmapSize(icon, sizeIcon, context)
    }

    fun createMaxSizeIcon(icon: Drawable, context: Context): Bitmap {
        return createIconBitmapSize(icon, sizeMax, context)
    }

    /**
     * Returns a bitmap suitable for the all apps view. The bitmap will be a
     * power of two sized ARGB_8888 bitmap that can be used as a gl texture.
     */
    private fun createIconBitmapSize(icon: Drawable, size: Int, context: Context): Bitmap {
        synchronized(sCanvas) {
            // we share the statics :-(
            if (sIconSize == -1) {
                initStatics(context)
            }

            var width: Int
            var height: Int

            when (size) {
                sizeIcon -> {
                    width = sIconSize
                    height = sIconSize
                }
                sizeMax -> {
                    width = sIconMaxScale
                    height = sIconMaxScale
                }
                else -> {
                    width = sIconSystem
                    height = sIconSystem
                }
            }

            if (icon is PaintDrawable) {
                icon.intrinsicWidth = width
                icon.intrinsicHeight = height
            } else if (icon is BitmapDrawable) {
                // Ensure the bitmap has a density.
                val bitmap = icon.bitmap

                if (bitmap.density == Bitmap.DENSITY_NONE) {
                    icon.setTargetDensity(context.resources.displayMetrics)
                }
            }
            val sourceWidth = icon.intrinsicWidth
            val sourceHeight = icon.intrinsicHeight

            if (sourceWidth > 0 && sourceWidth > 0) {
                // There are intrinsic sizes.
                if (width < sourceWidth || height < sourceHeight) {
                    // It's too big, scale it down.
                    val ratio = sourceWidth.toFloat() / sourceHeight
                    if (sourceWidth > sourceHeight) {
                        height = (width / ratio).toInt()
                    } else if (sourceHeight > sourceWidth) {
                        width = (height * ratio).toInt()
                    }
                } else if (width > sourceWidth && height > sourceHeight) {
                    // It's small, use the size they gave us.
                    width = sourceWidth
                    height = sourceHeight
                } else if (width == sourceWidth && height == sourceHeight) {
                    // no processing required
                    if (icon is BitmapDrawable) {
                        return icon.bitmap
                    }
                }
            }

            val bitmapSize = getSize(size, width, height)
            val bitmap = Bitmap.createBitmap(bitmapSize, bitmapSize, Bitmap.Config.ARGB_8888)
            val canvas = sCanvas
            canvas.setBitmap(bitmap)

            sOldBounds.set(icon.bounds)

            val left = (bitmapSize - width) / 2
            val top = (bitmapSize - height) / 2
            icon.setBounds(left, top, left + width, top + height)

            icon.colorFilter = null
            icon.draw(canvas)
            icon.bounds = sOldBounds
            return bitmap
        }
    }

    private fun getSize(size: Int, width: Int, height: Int): Int {
        return if (size == sizeMax && (width >= sIconMaxScale || height >= sIconMaxScale)) {
            sIconMaxScale
        } else if (size == sizeIcon && (width >= sIconSize || height >= sIconSize)) {
            sIconSize
        } else if (width > sIconSystem || height > sIconSystem) {
            if (width > height) width else height
        } else {
            sIconSystem
        }
    }

    fun canUseForInBitmap(candidate: Bitmap, targetOptions: BitmapFactory.Options): Boolean {
        // From Android 4.4 (KitKat) onward we can re-use if the byte size of
        // the new bitmap is smaller than the reusable bitmap candidate
        // allocation byte count.
        val width = targetOptions.outWidth / targetOptions.inSampleSize
        val height = targetOptions.outHeight / targetOptions.inSampleSize
        val byteCount = width * height * getBytesPerPixel(candidate.config)
        return byteCount <= candidate.allocationByteCount
    }

    private fun getBytesPerPixel(config: Bitmap.Config): Int {
        return when (config) {
            Bitmap.Config.ARGB_8888 -> 4
            Bitmap.Config.RGB_565 -> 2
            Bitmap.Config.ARGB_4444 -> 2
            Bitmap.Config.ALPHA_8 -> 1
            else -> 1
        }
    }

    fun flattenBitmap(bitmap: Bitmap): ByteArray? {
        // Try go guesstimate how much space the icon will take when serialized
        // to avoid unnecessary allocations/copies during the write.
        val size = bitmap.width * bitmap.height * 4
        val out = ByteArrayOutputStream(size)
        return try {
            bitmap.compress(Bitmap.CompressFormat.PNG, 100, out)
            out.flush()
            out.close()
            out.toByteArray()
        } catch (e: IOException) {
            Log.w("Favorite", "Could not write icon")
            null
        }

    }
}
